File: Data Structures & Algorithms/word-break/submission-0.cpp
class Solution {
public:
    unordered_map<int,bool>cache;
    bool dfs(string s,vector<string>& wordDict,int curr,int n){
        if(curr >= n){
            return true;
        }

        if(cache.find(curr) != cache.end()){
            return cache[curr];
        }

        bool ans = false;
        for(auto word: wordDict){
            int wordSize = word.size();
            string currWord = s.substr(curr,wordSize);

            if(currWord == word){
                ans |= dfs(s,wordDict,curr+wordSize,n);
            }
        }
        return cache[curr] = ans;
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        return dfs(s,wordDict,0,n);
    }
};

File: Data Structures & Algorithms/search-for-word/submission-0.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        m,n,k = len(board), len(board[0]), len(word)
        
        def dfs(i:int, j:int, idx: int) -> bool:
            if i < 0 or j < 0 or i >= m or j >= n or board[i][j] != word[idx]:
                return False
            
            if idx == k-1:
                return True
            
            ans = False
            tmp = board[i][j]
            board[i][j] = '-'

            ans |= dfs(i+1,j,idx+1)
            ans |= dfs(i-1,j,idx+1)
            ans |= dfs(i,j+1,idx+1)
            ans |= dfs(i,j-1,idx+1)

            board[i][j] = tmp
        
            return ans
        

        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:
                    if dfs(i,j,0):
                        return True
        
        return False
